apis/python/tests/test_basic_anndata_io.py

```diff
@@ -472,6 +473,108 @@ def test_add_matrix_to_collection(conftest_pbmc_small, tmp_path):
+    """Test shape validation for all matrix collection types."""
+    output_path = tmp_path.as_posix()
+    name, expected_shape = name_and_expected_shape
+    original_path = ROOT_DATA_DIR / "soma-experiment-versions-2025-04-04" / version / name
+    uri = str(original_path)
+    if not Path(uri).is_dir():
+        raise RuntimeError(
+            f"Missing '{uri}' directory. Try running `make data` from the TileDB-SOMA project root directory.",
+        )
+
+    shutil.copytree(uri, output_path, dirs_exist_ok=True)
+    n_obs = expected_shape[0]
+    n_vars = expected_shape[1]
+
+    # Determine expected shape based on matrix type
+    shape_map = {
+        "obsm": (n_obs, 5),
+        "varm": (n_vars, 5),
+        "obsp": (n_obs, n_obs),
+        "varp": (n_vars, n_vars),
+    }
+    base_shape = shape_map[matrix_type]
+    bad_shape = (max(1, base_shape[0] + offset[0]), max(1, base_shape[1] + offset[1]))
+
+    # Create invalid matrix
+    bad_matrix = csr_matrix((bad_shape[0], bad_shape[1])) if matrix_type in ["obsp", "varp"] else np.ones(bad_shape)
+
+    # Should raise error for invalid shape
+    if version.startswith("1.15"):
+        if offset != (0, 0):
+            with _factory.open(output_path, "w") as exp, pytest.raises(tiledbsoma.SOMAError):
+                tiledbsoma.io.add_matrix_to_collection(exp, "RNA", matrix_type, "bad", bad_matrix)
+        else:
+            with _factory.open(output_path, "w") as exp:
+                tiledbsoma.io.add_matrix_to_collection(exp, "RNA", matrix_type, "bad", bad_matrix)
+    else:
+        if offset == (0, 0):
+            with _factory.open(output_path, "w") as exp:
+                tiledbsoma.io.add_matrix_to_collection(exp, "RNA", matrix_type, "bad", bad_matrix)
+        else:
+            # TODO: _maybe_soma_joinid_shape returns max-domain for pre-1.14 instead of None
+            # https://github.com/single-cell-data/TileDB-SOMA/pull/4258#discussion_r2429676202
+            pass
+
+
+@pytest.mark.parametrize(
+    "offset",
+    [
+        (1, 0),
+        (0, 1),
+        (1, 1),
+    ],
+)
+def test_matrix_X_layer_enforcement(conftest_pbmc_small, tmp_path, offset):
+    """Test shape validation for all matrix collection types."""
+    output_path = tmp_path.as_posix()
+    original = conftest_pbmc_small.copy()
+
+    # Create SOMA experiment
+    tiledbsoma.io.from_anndata(output_path, conftest_pbmc_small, measurement_name="RNA")
+    assert_adata_equal(original, conftest_pbmc_small)
+
+    # Determine expected shape based on matrix type
+    shape_map = conftest_pbmc_small.n_obs, conftest_pbmc_small.n_vars
+    base_shape = shape_map
+    bad_shape = (max(1, base_shape[0] + offset[0]), max(1, base_shape[1] + offset[1]))
+
+    # Create invalid matrix
+    bad_matrix = np.ones(bad_shape)
+
+    # Should raise error for invalid shape
+    with _factory.open(output_path, "w") as exp, pytest.raises(tiledbsoma.SOMAError):
```

Comment:
nit: I would add a `match` arg to `pytest.raises`, just to confirm you are getting the Error message you expect.

E.g., `pytest.raises(tiledbsoma.SOMAError, match=r"Matrix .* shape")` or something like that....

Reply:
Added a regex match for the SOMAError.

Reply:
nit: actually, it looks like you did not add it - the code still lacks a `match` param.

Reply:
Oh I missed the second test, I had done it for the other one. Fixed now.